### ISSUE TITLE
Presence events do not report correctly for minions with NAT-ed IP

### ISSUE THREAD

**Issue by makearl** [saltstack/salt#26643]
The `salt/presence/present` event shows an empty list of present minions when you connect a minion that's behind a firewall with an IP that is NAT-ed. The presence event works correctly when the minion's IP is not NAT-ed.

Steps to recreate:
- Set up a salt master on AWS and add the following line to the master config:

```
presence_events: True
```
- Set up a salt minion on a network behind a firewall (if you set the minion up in AWS, you can point it to the external IP of the salt master to recreate the issue)
- Accept the salt minion's key on the master
- Listen for events on the master (`salt-run state.event pretty=True`). When you see presence events reported, present minions will be empty.

Salt master version information:

```
sudo salt --versions-report
                  Salt: 2015.5.3
                Python: 2.7.6 (default, Mar 22 2014, 22:59:56)
                Jinja2: 2.7.2
              M2Crypto: 0.21.1
        msgpack-python: 0.3.0
          msgpack-pure: Not Installed
              pycrypto: 2.6.1
               libnacl: Not Installed
                PyYAML: 3.10
                 ioflo: Not Installed
                 PyZMQ: 14.0.1
                  RAET: Not Installed
                   ZMQ: 4.0.4
                  Mako: Not Installed
               Tornado: Not Installed
 Debian source package: 2015.5.3+ds-1trusty1
```

**Comment by jfindlay** [comment]
@makearl, thanks for the report.

**Comment by makearl** [comment]
At the bottom of https://github.com/saltstack/salt/issues/18436 they touch on the NAT issue, but it was never resolved

**Comment by githubcdr** [comment]
Is there a way to debug this? I noticed that none of my Arch Linux salt-minions show up in presence events, but Ubuntu minions do, these minions are all behind a NAT firewall. Perhaps it's zeroMQ version related?

Running salt-minion -l all does not show any presence events being requested or send.

Is there an easy way to enable debug output for this?

**Comment by jfindlay** [comment]
@githubcdr, you can try setting `zmq_monitor: True` in the master and minion configs.

**Comment by githubcdr** [comment]
Hi @jfindlay,


I tried to debug this with zmq_monitor enabled but this does not reveal additional logging on the master or the minion related to presence events.

Is there another approach to debug this issue? From what I understand, it's the master sending the presence request. If this is true I could debug the function responsible for responding to this request, right?

**Comment by jfindlay** [comment]
@githubcdr, those are good questions, but this is beyond my current experience level in this area.  You may want to ask @cachedout or @jacksontj.

**Comment by paul-mulvihill** [comment]
Anyone made any headway on this? Would be really useful to use to show minions are connected.

**Comment by jacksontj** [comment]
I'm not *especially* up-to-date with presence events, but I've taken a look (and if needed we can go take a crack at fixing it).

I did some quick digging and it seems that this has a few potential problems (seems that the presence events have a few latent bugs ;) ). So to start off here is the code -- https://github.com/saltstack/salt/blob/develop/salt/utils/minions.py#L573 . 

From what I see there we have a couple problems:
1. No ipv6 support
2. lots of reliance on `data.p` (grains cache) -- seemingly if you have grains caching disabled you won't get anything
3. It is seemingly checking the minion IP address against the master IPs? (https://github.com/saltstack/salt/blob/develop/salt/utils/minions.py#L607)
    ```
    salt.utils.network.ip_addrs()
    ['10.0.1.116', '10.0.2.1', '172.17.0.1', '192.168.0.1', '192.168.122.1']
    ```

So because of that I spun some boxes up locally to test this, and it seems even more confusing. In my local testing I can't get the presence events to show anything besides the minion *on* the master (due to that `in addrs` check).

I did some quick testing of the presence events using tcp transport-- and those seem to be working just fine, so an option might be to switch?
